urldeb33fd2fa_friedemann_distripedia_tgz-pJ8-org_json_TestJ8/src/grinder/client/user/Browser.java

```diff
@@ -0,0 +1,242 @@
+		baseURIws = uriWS;
+
+		switch (type)
+		{
+		case WEBSOCKET:
+			LS = new LocalStorage();
+			dpc = new DistripediaClient(uriWS, LS);
+			dpc.setCachedArticlesPreferred(true);
+			dpc.addArticleReceivedHandler(this);
+			break;
+
+		case STRATUS:
+			LS = new LocalStorage();
+			break;
+		}
+	}
+
+	/**
+	 * Gets the browser id.
+	 *
+	 * @return the browser id
+	 */
+	public UUID getBrowserID() {
+		return browserID;
+	}
+
+	/**
+	 * Gets the browser type.
+	 *
+	 * @return BrowserType the browser type
+	 * @see BrowserType
+	 */
+	public BrowserType getBrowserType() {
+		return type;
+	}
+
+	/*
+	 * (non-Javadoc)
+	 * @see
+	 * websocket.client.DistripediaClient.ArticleReceivedHandler#onArticleReceived
+	 * (websocket.client.DistripediaClient.ArticleReceivedEvent)
+	 */
+	@Override
+	public void onArticleReceived(final ArticleReceivedEvent e) {
+		articleReceivedEvent = e;
+
+		// new (david)
+		synchronized (waitingForArticleLock)
+		{
+			waitingForArticleLock.notify();
+		}
+		// end new
+	}
+
+	public ArticleRequest retrieveArticle(final String articleID) {
+
+		final ArticleRequest ar = new ArticleRequest();
+
+		switch (type)
+		{
+		case HTTP:
+
+			retrieveArticleViaHTTP(ar, baseURI.toString() + articleID + ".html");
+			break;
+
+		case WEBSOCKET:
+			articleReceivedEvent = null;
+
+			dpc.getArticle(articleID);
+
+			// new (david)
+			try
+			{
+				synchronized (waitingForArticleLock)
+				{
+					waitingForArticleLock
+							.wait(GrinderSettings.CLIENT_WEBSOCKET_RECEIVE_TIMEOUT);
+				}
+			} catch (final InterruptedException e)
+			{}
+			// end new
+
+			// final long endTime = System.currentTimeMillis() +
+			// GrinderSettings.CLIENT_WEBSOCKET_RECEIVE_TIMEOUT;
+			// while (!hasArticleReceived)
+			// {
+			//
+			// if (endTime < System.currentTimeMillis())
+			// {
+			// break;
+			// }
+			// }
+
+			if (articleReceivedEvent != null)
+			{
+				final int articleSize = articleReceivedEvent.getContent().length();
+				final String articleTitle = articleReceivedEvent.getArticle();
+
+				ar.articleIdReceived = articleTitle;
+				ar.bytesReceived = articleSize;
+			} else
+			{
+				ar.articleIdReceived = "TIMEOUT";
+				ar.bytesReceived = -1;
+			}
+			break;
+
+		case STRATUS:
+			// TODO implement stratus emulating browser connections
+
+			// get browserid from server
+			// setup connection
+			// receive article from peer user
+			// close connection
+
+			break;
+		}
+
+		return ar;
+	}
+
+	/**
+	 * Retrieve article(-size) via http.
+	 *
+	 * @param address the address
+	 * @return int bytes received
+	 */
+	public ArticleRequest retrieveArticleViaHTTP(final ArticleRequest ar,
+			final String address) {
+		URL url = null;
+		try
+		{
+			url = new URL(address);
+		} catch (final MalformedURLException e1)
+		{
+			// TODO Auto-generated catch block
+			e1.printStackTrace();
+		}
+		String html = "";
+
+		try
+		{
+			final BufferedReader br = new BufferedReader(new InputStreamReader(url
+					.openStream()));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the resource created by the following constructor: *InputStreamReader*. Execution paths that do not contain closure statements were detected. Ensure the object returned by *InputStreamReader()* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).